src/pyop/provider.py

```diff
@@ -339,7 +340,9 @@ def _do_code_exchange(self, request,  # type: Dict[str, str]
             raise InvalidTokenRequest(str(e), token_request) from e
 
         authentication_request = self.authz_state.get_authorization_request_for_code(token_request['code'])
-
+        
+        if token_request['client_id'] != authentication_request['client_id']:
```

Comment:
Should probably log this case, since it might indicate some security issues on the RP side:
`logger.info('Authorization code \'%s\' belonging to \'%s\' was used by \'%s\'', token_request['code'], authentication_request['client_id'], token_request['client_id'])`